ArcGISRuntimeSDKQt_CppSamples/Maps/DownloadPreplannedMap/DownloadPreplannedMap.cpp

```diff
@@ -0,0 +1,243 @@
+  qmlRegisterType<DownloadPreplannedMap>("Esri.Samples", 1, 0, "DownloadPreplannedMapSample");
+}
+
+MapQuickView* DownloadPreplannedMap::mapView() const
+{
+  return m_mapView;
+}
+
+// Set the view (created in QML)
+void DownloadPreplannedMap::setMapView(MapQuickView* mapView)
+{
+  if (!mapView || mapView == m_mapView)
+    return;
+
+  m_mapView = mapView;
+  m_mapView->setMap(m_map);
+
+  emit mapViewChanged();
+}
+
+void DownloadPreplannedMap::checkIfMapAreaIsLoaded(int index)
+{
+  if (!m_offlineMapTask)
+    return;
+
+  if (m_offlineMapTask->preplannedMapAreaList()->isEmpty())
+    return;
+
+  PreplannedMapArea* mapArea = m_offlineMapTask->preplannedMapAreaList()->at(index);
+
+  if (!mapArea || mapArea->loadStatus() != LoadStatus::Loaded)
+    return;
+
+  m_busy = true;
+  emit busyChanged();
+
+  loadSelectedMap(*mapArea, index);
+}
+
+void DownloadPreplannedMap::loadSelectedMap(const PreplannedMapArea& mapArea, int index)
+{
+  const QString path = QString(m_tempPath.path() + "/" + mapArea.portalItem()->title());
+
+  QDir tempDir(path);
+
+  if (m_preplannedMapExists)
+    loadExistingPreplannedMap(path);
+  else
+    downloadPreplannedMapArea(path, index);
+}
+
+void DownloadPreplannedMap::checkIfMapExists(int index)
+{
+  if (!m_offlineMapTask)
+    return;
+
+  if (m_offlineMapTask->preplannedMapAreaList()->isEmpty())
+    return;
+
+  PreplannedMapArea* mapArea = m_offlineMapTask->preplannedMapAreaList()->at(index);
+
+  if (!mapArea || mapArea->loadStatus() != LoadStatus::Loaded)
+    return;
+
+  const QString path = QString(m_tempPath.path() + "/" + mapArea->portalItem()->title());
+  QDir tempDir(path);
+
+  m_preplannedMapExists = tempDir.exists();
+  emit preplannedMapExistsChanged();
+
+  if (m_viewingOnlineMaps)
+    m_mapView->setViewpointGeometry(mapArea->areaOfInterest());
+}
+
+void DownloadPreplannedMap::showOnlineMap(int index)
+{
+  m_viewingOnlineMaps = true;
+  emit viewingOnlineMapsChanged();
+  m_mapView->setMap(m_map);
+  const Geometry areaOfInterest = m_offlineMapTask->preplannedMapAreaList()->at(index)->areaOfInterest();
+  m_mapView->setViewpointGeometry(areaOfInterest);
+  m_graphicsOverlay->setVisible(true);
+}
+
+void DownloadPreplannedMap::onDownloadPreplannedMapJobCompleted()
+{
+  if (m_preplannedMapJob->jobStatus() != JobStatus::Succeeded)
+    return;
+
+  m_mapView->setMap(m_preplannedMapJob->result()->offlineMap());
+  m_offlineMapTask->disconnect();
+  m_busy = false;
+  emit busyChanged();
+  m_preplannedMapExists = true;
+  emit preplannedMapExistsChanged();
+  m_viewingOnlineMaps = false;
+  emit viewingOnlineMapsChanged();
+  m_graphicsOverlay->setVisible(false);
+}
+
+void DownloadPreplannedMap::loadPreplannedMapAreas()
+{
+  m_preplannedList = m_offlineMapTask->preplannedMapAreaList();
+  m_busy = false;
+  emit preplannedListChanged();
+  emit busyChanged();
+
+  for (PreplannedMapArea* mapArea : *static_cast<PreplannedMapAreaListModel*>(m_preplannedList))
```

Comment:
Probably is better to pre-cache `*static_cast<PreplannedMapAreaListModel*>(m_preplannedList)` before the loop